consensus/state.go

```diff
@@ -1267,7 +1267,21 @@ func (cs *State) defaultDecideProposal(height int64, round int32) {
 	// Decide on block
 	if cs.ValidBlock != nil {
 		// If there is valid block, choose that.
-		block, blockParts = cs.ValidBlock, cs.ValidBlockParts
+		block = cs.ValidBlock
+
+		// set the recovery related fields if using an existing block
+		hashes := make([][]byte, len(block.Txs))
+		for i := 0; i < len(block.Txs); i++ {
+			hashes[i] = block.Txs[i].Hash()
+		}
+		block.SetCachedHashes(hashes)
+
+		parts, err := block.MakePartSet(types.BlockPartSizeBytes)
```

Comment:
Is it always required to `MakePartSet`? I don't have a full context, but previously `cs.ValidBlockParts` were returned, so I assume it worked correctly at least sometimes ;)

Reply:
With the propagation reactor it is. In fact, the `MakePartSet` method is updated, with the recovery work, to scan the protobuf encoded block to find the transactions positions. However, we didn't want to change a lot more in the consensus reactor, so we didn't make it a rule to also do the same whenever we're constructing the partset. So, at this level, we have the partset, but we don't have the transactions positions, which is expected by the propagation reactor. So, we manually populate it here.

Without the propagation reactor, you're right, recreating the partset is totally unnecessary and `cs.ValidBlockParts` has the expect value.